### ISSUE TITLE
ElementalArea GridField does not support Blocks with same name but different namespace

### ISSUE THREAD

**Issue by chrispenny** [dnadesign/silverstripe-elemental#751]
## Affects
4.2.1
4.3.0

## Problem

If your project namespaces by feature (rather than namespacing by type), it's expected that your Block class names are the same, but that they live under a different namespace.

It doesn't seem like there is support for this, as any time you create a new Block, you get the same "Untitled [first alphabetical block name] block", and also, every Block in the Gridfield has the same icon. **Also**, if you have a mix of `inline_editable`, you will also get whatever config was set for the first block.

Pictured below are two different Block types. They each have a different value being returned for `getType()`, and each has a different `$icon` defined.

![Screen Shot 2019-11-15 at 11 01 59 AM](https://user-images.githubusercontent.com/505788/68899835-6ee24480-0797-11ea-85a0-85afabd982d0.png)

I then tried to add a new Block, but with a unique namespace **and** unique class name, and things works correctly for that Block only.

![Screen Shot 2019-11-15 at 12 05 40 PM](https://user-images.githubusercontent.com/505788/68903614-91c52680-07a0-11ea-8d3d-02d71c25a949.png)

### Responses when creating my second Block (with the same class name)

TL;DR: `typeName` for both is `AppBlock`.

```
{
  "data": {
    "addElementToArea": {
      "ID": "2",
      "__typename": "Block"
    }
  }
}
```

```
{
  "data": {
    "readOneElementalArea": {
      "Elements": [
        {
          "ID": "2",
          "Title": null,
          "BlockSchema": {
            "typeName": "AppBlock",
            "actions": {
              "edit": "http://xx.local/admin/pages/edit/show/1"
            }
          },
          "IsLiveVersion": false,
          "IsPublished": false,
          "Version": 1,
          "__typename": "Block"
        },
        {
          "ID": "1",
          "Title": null,
          "BlockSchema": {
            "typeName": "AppBlock",
            "actions": {
              "edit": "http://xx.local/admin/pages/edit/show/1"
            }
          },
          "IsLiveVersion": false,
          "IsPublished": false,
          "Version": 1,
          "__typename": "Block"
        }
      ],
      "__typename": "ElementalArea"
    }
  }
}
```

### Responses when creating my third Block (with the different class name)

TL;DR: `typeName` for new Block is `AppIconWallBlock`, while the other two are still `AppBlock`.

```
{
  "data": {
    "addElementToArea": {
      "ID": "3",
      "__typename": "Block"
    }
  }
}
```

```
{
  "data": {
    "readOneElementalArea": {
      "Elements": [
        {
          "ID": "3",
          "Title": null,
          "BlockSchema": {
            "typeName": "AppIconWallBlock",
            "actions": {
              "edit": "http://xx.local/admin/pages/edit/show/1"
            }
          },
          "IsLiveVersion": false,
          "IsPublished": false,
          "Version": 1,
          "__typename": "Block"
        },
        {
          "ID": "2",
          "Title": null,
          "BlockSchema": {
            "typeName": "AppBlock",
            "actions": {
              "edit": "http://xx.local/admin/pages/edit/show/1"
            }
          },
          "IsLiveVersion": false,
          "IsPublished": false,
          "Version": 1,
          "__typename": "Block"
        },
        {
          "ID": "1",
          "Title": null,
          "BlockSchema": {
            "typeName": "AppBlock",
            "actions": {
              "edit": "http://xx.local/admin/pages/edit/show/1"
            }
          },
          "IsLiveVersion": false,
          "IsPublished": false,
          "Version": 1,
          "__typename": "Block"
        }
      ],
      "__typename": "ElementalArea"
    }
  }
}
```

### BaseElement::provideBlockSchema()

`StaticSchema::inst()->typeNameForDataObject(static::class)` looks to simply grab the last part of the class name, and this is being returned for `typeName`.

### ElementTypeRegistry::registerElement()

Is also using `StaticSchema::inst()->typeNameForDataObject($elementClass)` for `$typeName`.

### Hero banner block

```php
<?php

namespace App\Block\HeroBanner;

use DNADesign\Elemental\Models\BaseElement;
use SilverStripe\Assets\Image;

class Block extends BaseElement
{
    private static $icon = 'font-icon-block-banner';

    private static $table_name = 'HeroBannerBlock';

    private static $db = [
        'Content' => 'Varchar(255)',
    ];

    private static $has_one = [
        'DesktopImage' => Image::class,
        'MobileImage' => Image::class,
    ];

    public function getType(): string
    {
        return 'Hero Banner';
    }
}
```

### Text block

```php
<?php

namespace App\Block\Text;

use DNADesign\Elemental\Models\BaseElement;

class Block extends BaseElement
{
    private static $icon = 'font-icon-block-content';

    private static $db = [
        'HTML' => 'HTMLText',
    ];

    private static $table_name = 'TextBlock';

    public function getType(): string
    {
        return 'Text';
    }
}
```

### Icon wall block

```
<?php

namespace App\Block\IconWall;

use DNADesign\Elemental\Models\BaseElement;

class IconWallBlock extends BaseElement
{
    private static $icon = 'font-icon-block-carousel';

    private static $table_name = 'IconWallBlock';

    private static $description = 'icon wall';

    public function getType(): string
    {
        return 'IconWall';
    }
}
```

## Why namespace by feature?

When your project gets to a certain size, it becomes difficult to "find all of the pieces" if you have namespaced by "type", EG:
```
App /
    - Model /
        - Block /
            - ...
    - Service /
        - Hydration /
            - ...
        - Result /
            - ...
    ...
```

Vs namespacing by feature, EG:
```
App /
    - Element /
        - Text /
            - Block
            - HydrationService
            - Result
        - Carousel /
            - Block
            - HydrationService
            - Result
    ...
```

**Comment by ScopeyNZ** [comment]
This is related to the default "type name" resolved by the GraphQL module. You can define custom type names as per the GraphQL docs:

https://github.com/silverstripe/silverstripe-graphql#customising-the-names-of-types-and-operations

This _might_ resolve the issue, not quite sure.

**Comment by ScopeyNZ** [comment]
I've omitted a `change/` label as I'm not sure this can actually be fixed in a patch release if there is a bug here.

**Comment by chrispenny** [comment]
Thanks, @ScopeyNZ !

I gave this a go:
```yml
SilverStripe\GraphQL\Manager:
  schemas:
    default:
      typeNames:
        App\Block\Text\Block: TextBlog
        App\Block\HeroBanner\Block: HeroBanner
```

But there appears to be no change.

When I updated `BaseElement::provideBlockSchema()` and `ElementTypeRegistry::registerElement()` to stop using `StaticSchema::inst()->typeNameForDataObject( ... name ...)`, things begin to work.

```php
return [
    // Currently GraphQL doesn't expose the correct type name and just returns "base element"s. This is a
    // workaround until we can scaffold a query client side that specifies by type name
    'typeName' => StaticSchema::inst()->typeNameForDataObject(static::class),
    'actions' => [
        'edit' => $this->getEditLink(),
    ],
];
```

**Comment by unclecheese** [comment]
```
SilverStripe\GraphQL\Manager:
  schemas:
    default:
```

Should be applied to the `admin` schema:

```
SilverStripe\GraphQL\Manager:
  schemas:
    admin:
```

Can you try that?

**Comment by chrispenny** [comment]
That's done the trick! Thanks, @unclecheese !

If this is "working as intended", then I'm happy to close this, and probably I'll just add some documentation.

**Comment by chrispenny** [comment]
@unclecheese sorry to resurrect this, but, actually I'm not sure this is working.

I've just come back to it, and realised that (and I don't understand why), it only works when a `flush=1` is present.

## With flush

![Screen Shot 2019-12-04 at 2 13 29 PM](https://user-images.githubusercontent.com/505788/70103563-55e6f800-16a0-11ea-89f4-761deb390e0c.png)

## Without flush

![Screen Shot 2019-12-04 at 2 13 40 PM](https://user-images.githubusercontent.com/505788/70103582-64351400-16a0-11ea-9bc8-9444a1148ce3.png)

Console shows the following error:
```
Uncaught TypeError: Cannot read property 'title' of undefined
    at Element.render (Element.js:192)
```

Which references:
```
    const linkTitle = i18n.inject(
      i18n._t('ElementalElement.TITLE', 'Edit this {type} block'),
      { type: type.title }
    );
```

Why it works with `flush`, I don't know... Project code is the same basic set up as before.

The solution/question I had above, still seems to solve the issue:
```
// BaseElement::provideBlockSchema()
'typeName' => str_replace('\\', '', static::class),

// ElementTypeRegistry::registerElement()
$typeName = str_replace('\\', '', $elementClass)
```

**Comment by ScopeyNZ** [comment]
Hmmmn. There's a cache for element types - do you run on multiple web servers?

**Comment by chrispenny** [comment]
@ScopeyNZ This is just on my local machine.

**Comment by chrispenny** [comment]
Looks like my original "solution" isn't working either. Though I have updated to version `4.3.0`.

**Comment by brynwhyman** [comment]
Side note: If we're going to harden the codebase to support this approach, it seems like it might be worth a mention in the documentation that it's possible and available (presumably for larger sitges) to namespace your blocks by 'feature'

**Comment by chrispenny** [comment]
Hi team,

I just wanted to add an additional note:
I had been playing around with `silverstripe-gatsby` recently, and noticed that it had a similar issue (this time it wasn't with `Element` objects though). I also tried updating the `typeNames` for that using the config method/s described above, and it similarly did not work. Perhaps the bug is in our graphql module?

**Comment by ScopeyNZ** [comment]
cc @unclecheese

**Comment by unclecheese** [comment]
Looked into this. The issue is that `typeNames ` are only set in the context of a GraphQL request, when the `Manager` config is booted. If you need access to typenames in any other context, e.g. a `getCMSFields()` call, they will be empty.

Quick fix:

```php
$schemaConfig = Manager::config()->get('schemas');
$typeNames = $schemaConfig['admin']['typeNames'] ?? [];
StaticSchema::inst()->setTypeNames($typeNames);
$typeName = StaticSchema::inst()->typeNameForDataObject(MyObject::class);
```

Long-term fix: https://github.com/silverstripe/silverstripe-graphql/pull/254

**Comment by mfendeksilverstripe** [comment]
I can confirm that this issue was resolved by https://github.com/silverstripe/silverstripe-graphql/pull/254

The fix was released as a part `silverstripe/silverstripe-graphql` `3.3.0` release.

## Test

![Screen Shot 2020-08-31 at 9 58 52 AM](https://user-images.githubusercontent.com/26395487/91670364-e2abdd00-eb70-11ea-9e32-5a13cd1869ff.png)

![Screen Shot 2020-08-31 at 9 58 58 AM](https://user-images.githubusercontent.com/26395487/91670368-e63f6400-eb70-11ea-893f-6fc5e1f8e594.png)

Note that it's necessary to properly configure the blocks to display the right lables.

## Example

```
<?php

namespace App\Elemental\Apple;

use DNADesign\Elemental\Models\BaseElement;
use SilverStripe\Forms\FieldList;
use SilverStripe\Forms\TextField;

class Block extends BaseElement
{
    /**
     * @var string
     */
    private static $icon = 'font-icon-p-list';

    /**
     * @var string
     */
    private static $table_name = 'AppleBlock';

    /**
     * @var string
     */
    private static $singular_name = 'Apple block';

    /**
     * @var string
     */
    private static $plural_name = 'Apple blocks';

    /**
     * @var string
     */
    private static $description = 'Block with apples';

    /**
     * @var array
     */
    private static $db = [
        'Colour' => 'Varchar',
    ];

    public function getType(): string
    {
        return $this->i18n_singular_name();
    }

    public function getCMSFields(): FieldList
    {
        $fields = parent::getCMSFields();
        $fields->addFieldsToTab('Root.Main', [
            TextField::create('Colour'),
        ]);

        return $fields;
    }
}

```

**Comment by mfendeksilverstripe** [comment]
@chrispenny This issue can be closed now.
